mermithid/processors/misc/ComplexLineShape.py

```diff
@@ -0,0 +1,598 @@
+
+# Takes only the nonzero bins of a histogram
+def get_only_nonzero_bins(bins,hist):
+    nonzero_idx = np.where(hist!=0)
+    hist_nonzero = hist[nonzero_idx]
+    hist_err = np.sqrt(hist_nonzero)
+    bins_nonzero = bins[nonzero_idx]
+    return bins_nonzero , hist_nonzero , hist_err
+
+# Flips an array left-to-right. Useful for converting between energy and frequency
+def flip_array(array):
+    flipped = np.fliplr([array]).flatten()
+    return flipped
+
+# Given energy in keV and the self.B_field of the trap, returns frequency in Hz
+def energy_to_frequency(energy_vec, B_field):
+    freq_vec = (e_charge*B_field/((2.*np.pi*m_e)*(1+energy_vec/mass_energy_electron)))
+    return freq_vec
+
+# Given frequency in Hz and the self.B_field of the trap, returns energy in keV
+def frequency_to_energy(freq_vec,B_field):
+    energy_vec = (e_charge*B_field/((2.*np.pi*m_e*freq_vec))-1)*mass_energy_electron
+    return energy_vec
+
+# Converts an energy to frequency using a guess for magnetic field. Can handle errors too
+def energy_guess_to_frequency(energy_guess,energy_guess_err,B_field_guess):
+    frequency = energy_to_frequency(energy_guess,B_field_guess)
+    const = e_charge*B_field_guess/(2.*np.pi*m_e)
+    frequency_err = const/(1+energy_guess/mass_energy_electron)**2*energy_guess_err/mass_energy_electron
+    return frequency , frequency_err
+
+# Given a frequency and error, converts those to B field values assuming the line is the 17.8 keV line
+def central_frequency_to_B_field(central_freq,central_freq_err):
+    const = (2.*np.pi*m_e)*(1+kr_line/mass_energy_electron)/e_charge
+    B_field = const*central_freq
+    B_field_err = const*central_freq_err
+    return B_field , B_field_err
+
+# given a FWHM for the lorentian component and the FWHM for the gaussian component,
+# this function estimates the FWHM of the resulting voigt distribution
+def FWHM_voigt(FWHM_L,FWHM_G):
+    FWHM_V = 0.5346*FWHM_L+np.sqrt(0.2166*FWHM_L**2+FWHM_G**2)
+    return FWHM_V
+
+# Returns the name of the current directory
+def get_current_dir():
+    current_path = os.path.abspath(os.getcwd())
+    stripped_path = [s.strip('\n') for s in current_path]
+    stripped_path = stripped_path[0].split('/')
+    current_dir = stripped_path[len(stripped_path)-1]
+    return current_dir
+
+class ShakeSpectrumClass():
+    # The shakeup/shakeoff spectrum for the 17.8 keV line of Kr83m based on Hamish Vadantha paper
+    # Overleaf link for the paper https://www.overleaf.com/project/5d56e015162d244b1c283c57
+    #######################################################################################
+    # Read parameters for shake up shake off spectrum from an excel spread sheet.
+    # The parameters are from Hamish and Vedantha shake spectrum paper table IV
+
+    def __init__(self, shake_spectrum_parameters_json_path, input_std_eV_array):
+
+        with open(
```

Comment:
can you combine these lines? I don't think the statement is unreasonably long that it needs to be broken up. dont think it would be the longest in the file